sympy/matrices/matrices.py

```diff
@@ -2044,6 +2192,425 @@ def minor_submatrix(self, i, j):
+            normalized (so that the pivots are 1), or whether
+            rows should be normalized along the way (like the naive
+            row reduction algorithm)
+        normalize : whether pivot rows should be normalized so that
+            the pivot value is 1
+        zero_above : whether entries above the pivot should be zeroed.
+            If `zero_above=False`, an echelon matrix will be returned.
+        """
+        rows, cols = self.rows, self.cols
+        mat = list(self)
+        def get_col(i):
+            return mat[i::cols]
+
+        def row_swap(i, j):
+            mat[i*cols:(i + 1)*cols], mat[j*cols:(j + 1)*cols] = \
+                mat[j*cols:(j + 1)*cols], mat[i*cols:(i + 1)*cols]
+
+        def cross_cancel(a, i, b, j):
+            """Does the row op row[i] = a*row[i] - b*row[j]"""
+            q = (j - i)*cols
+            for p in range(i*cols, (i + 1)*cols):
+                mat[p] = a*mat[p] - b*mat[p + q]
+
+        piv_row, piv_col = 0, 0
+        pivot_cols = []
+        swaps = []
+        # use a fraction free method to zero above and below each pivot
+        while piv_col < cols and piv_row < rows:
+            pivot_offset, pivot_val, \
+            assumed_nonzero, newly_determined = _find_reasonable_pivot(
+                get_col(piv_col)[piv_row:], iszerofunc, simpfunc)
+
+            # _find_reasonable_pivot may have simplified some things
+            # in the process.  Let's not let them go to waste
+            for (offset, val) in newly_determined:
+                offset += piv_row
+                mat[offset*cols + piv_col] = val
+
+            if pivot_offset is None:
+                piv_col += 1
+                continue
+
+            pivot_cols.append(piv_col)
+            if pivot_offset != 0:
+                row_swap(piv_row, pivot_offset + piv_row)
+                swaps.append((piv_row, pivot_offset + piv_row))
+
+            # if we aren't normalizing last, we normalize
+            # before we zero the other rows
+            if normalize_last is False:
+                i, j = piv_row, piv_col
+                mat[i*cols + j] = S.One
+                for p in range(i*cols + j + 1, (i + 1)*cols):
+                    mat[p] = mat[p] / pivot_val
+                # after normalizing, the pivot value is 1
+                pivot_val = S.One
+
+            # zero above and below the pivot
+            for row in range(rows):
+                # don't zero our current row
+                if row == piv_row:
+                    continue
+                # don't zero above the pivot unless we're told.
+                if zero_above is False and row < piv_row:
+                    continue
+                # if we're already a zero, don't do anything
+                val = mat[row*cols + piv_col]
+                if iszerofunc(val):
+                    continue
+
+                cross_cancel(pivot_val, row, val, piv_row)
```

Comment:
Looks like you can get the same result with fewer calculations. When forming the linear combination of the pivot row with a row beneath it, it looks like you're explicitly computing the entry in the pivot column of the result, but this entry is zero by construction. When the pivot column is not zero, it looks like you're computing the entries to the left of the pivot in the linear combination of rows, but these are also zero by construction.